File: 112.PathSum/clywin123/hasPathSum.py
class Solution:
    # @param {TreeNode} root
    # @param {integer} sum
    # @return {boolean}
    def hasPathSum(self, root, sum):
        def _hasPathSum(root,sum,value):
            if(not root):
                return False
            value += root.val
            if(root.left==None and root.right==None):
                if(value == sum):
                    return True
                else:
                    return False
            else:
                return _hasPathSum(root.left,sum,value) or _hasPathSum(root.right,sum,value)
        return _hasPathSum(root,sum,0) 

File: 112.PathSum/wyrj/hasPathSum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {boolean}
 */
var hasPathSum = function(root, sum) {
    if (!root) return false;
    const remain = sum - root.val;
    if (!root.left && !root.right) return remain === 0;
    return hasPathSum(root.left, remain) || hasPathSum(root.right, remain);
};
